codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteAssigner.java

```diff
@@ -0,0 +1,437 @@
+        boolean hasFrame;
+        int streamableBytes;
+        int assigned;
+        int unassignedStreamableBytesForTree;
+
+        PriorityState(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        /**
+         * The stream this state is associated with.
+         */
+        Http2Stream stream() {
+            return stream;
+        }
+
+        /**
+         * Recursively increments the {@link #unassignedStreamableBytesForTree()} for this branch in
+         * the priority tree starting at the current node.
+         */
+        void unassignedStreamableBytesForTreeChanged(int delta) {
+            unassignedStreamableBytesForTree += delta;
+            if (!stream.isRoot()) {
+                state(stream.parent()).unassignedStreamableBytesForTreeChanged(delta);
+            }
+        }
+
+        void assign(int bytes) {
+            assigned += bytes;
+            totalAssigned += bytes;
+
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored before the assigned bytes are
+            // actually written.
+            unassignedStreamableBytesForTreeChanged(-bytes);
+        }
+
+        /**
+         * Reset the number of bytes that have been assigned to this stream by the priority
+         * algorithm.
+         */
+        void resetAssigned() {
+            assign(-assigned);
+        }
+
+        void update(StreamState streamState) {
+            hasFrame = streamState.hasFrame();
+
+            int newStreamableBytes = streamState.streamableBytes();
+
+            int delta = newStreamableBytes - streamableBytes;
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unassignedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unassign all bytes.
+            resetAssigned();
+
+            // Clear the streamable bytes.
+            update(new StreamState() {
+                @Override
+                public Http2Stream stream() {
+                    return stream;
+                }
+
+                @Override
+                public int streamableBytes() {
+                    return 0;
+                }
+
+                @Override
+                public boolean hasFrame() {
+                    return false;
+                }
+            });
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unassignedStreamableBytes() {
+            return streamableBytes - assigned;
+        }
+
+        int unassignedStreamableBytesForTree() {
+            return unassignedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class StreamVisitor implements Http2StreamVisitor {
```

Comment:
Maybe (probably) ... from my experience in Netty the general preference is to be explicit. I'm not strong on this ... @normanmaurer - WDYT?